bib_rights/gen_bib_rights_tests.pl

```diff
@@ -0,0 +1,1795 @@
+  ['coo.31924077881971', 'check for tech report number formats'],
+  ['mdp.39015024459466', '<space>NNNN-N<space>'],
+];
+
+
+my $test_struct = {};
+my $fixtures_dir = $ENV{'SDRROOT'} . '/crms/t/fixtures/bib_rights/';
+unless (-e $fixtures_dir) {
+  `mkdir -p $fixtures_dir`;
+}
+
+# Create fake fed doc OCLC numbers file (only used in the last coverage tests)
+my $oclc_file = $fixtures_dir . 'feddoc_oclc_nums.txt';
+open my $fh, '>:utf8', $oclc_file or die "error opening $oclc_file: $!";
+print $fh "0\n00\n000\n";
+close $fh;
+
+foreach my $test (@$tests) {
+  my ($htid, $comment) = @$test;
+  my $marc;
+  my $xml;
+  unless (defined $test_struct->{$htid}) {
+    $test_struct->{$htid} = {id => $htid, comment => $comment, years => {}};
+  }
+  my $fixture = $fixtures_dir . $htid . ".xml";
+  if (-e $fixture) {
+    open my $fh, '<:utf8', $fixture or die "error opening $fixture: $!";
+    $xml = do { local $/; <$fh> };
+    close $fh;
+  } else {
+    my $record = $crms->GetMetadata($htid);
+    if (!defined $record) {
+      die "Unable to get metadata for $htid\n";
+    }
+    $xml = $record->xml;
+    open my $fh, '>:utf8', $fixture or die "error opening $fixture: $!";
+    print $fh $xml;
+    close $fh;
+  }
+  eval { $marc = MARC::Record->new_from_xml($xml); };
+  if ($@) {
+    print STDERR "problem processing marc xml\n";
+    warn $@;
+    print STDERR "$xml\n";
+    next;
+  };
+  foreach my $year ((2010, 2015, 2020, 2025, 2030)) {
+    $ENV{BIB_RIGHTS_DATE} = $year;
+    my $br = bib_rights->new();
+    my $cid = $marc->field('001')->as_string();
+    my $enumcron = extract_enumcron($marc, $htid);
+    my $bib_info = $br->get_bib_info($marc, $cid);
+    my $bri = $br->get_bib_rights_info($htid, $bib_info, $enumcron);
+    $test_struct->{$htid}->{years}->{$year} = $bri->{attr};
+  }
+}
+
+write_json($test_struct, 'bib_rights_tests.json');
+gen_clean_pub_place_tests();
+
+# Extract enumcron for HTID from MARC::Record 974z
+sub extract_enumcron {
+  my $marc = shift;
+  my $htid = shift;
+  
+  my @fields = $marc->field('974');
+  foreach my $field (@fields) {
+    if ($htid eq $field->subfield('u')) {
+      return $field->subfield('z') || '';
+    }
+  }
+  return '';
+}
+
+# Write a structure to the fixtures directory.
+sub write_json {
+  my $struct = shift;
+  my $file = shift;
+
+  my $jsonxs = JSON::XS->new->utf8->canonical(1)->pretty(1);
+  my $encdata = $jsonxs->encode($struct);
+  my $fixture_file = $fixtures_dir . $file;
+  open $fh, '>:utf8', $fixture_file or die "error opening $fixture_file: $!";
+  print $fh $encdata;
+  close $fh;
+}
+
+# Tests for normalization of all the random garbage that shows up in 008 pub_place fields.
+# These are all the attested values and what bib_rights::clean_pub_place should produce.
+# SELECT DISTINCT CONCAT("['",SUBSTR(f008,16,3),"','",pub_place,"'],") AS s FROM bib_rights_bi ORDER BY s;
+sub gen_clean_pub_place_tests {
```

Comment:
Re clean_pub_place: everything is included there because it's attested. We could prolly whittle the tests down to those relevant to the US vs Commonwealth copyright term distinction. But that would complicate the test generator more than just testing everything we see in the catalog. I'm also not sure if we want to try repairing more of the noisy data we see there, if it's possible to do (there's a specific carve-out for Puerto Rico on bib_rights.pm line ~487, for example).

Regardless, the FIXME in t/bib_rights.t can go away because it's fixed.

And speaking of the generator. That's meant to stay in CRMS because it's using CRMS infrastructure to bootstrap the fixtures (spefically its own Bib API setup as well as the bib_rights_bi table distilled from the catalog). The JSON is the bit that is intended to go into post_zephir_processing. The data in `gen_clean_pub_place_tests` is copied from a MariaDB query. So -- alas -- nothing's truly canonical here.